core/src/apps/cardano/sign_tx.py

```diff
@@ -245,7 +391,24 @@ async def _show_tx(
     ctx: wire.Context, keychain: seed.Keychain, msg: CardanoSignTx
 ) -> None:
     total_amount = await _show_outputs(ctx, keychain, msg)
-    await confirm_transaction(ctx, total_amount, msg.fee, msg.protocol_magic)
+
+    for certificate in msg.certificates:
+        await confirm_certificate(ctx, certificate)
+
+    for withdrawal in msg.withdrawals:
+        await confirm_withdrawal(ctx, withdrawal)
+
+    if msg.metadata_hash:
+        await confirm_metadata_hash(ctx, msg.metadata_hash)
+
+    await confirm_transaction(
+        ctx,
+        total_amount,
+        msg.fee,
+        msg.protocol_magic,
+        msg.certificates,
```

Comment:
So why do we have separate confirm_certificate and also pass certificates here?

Reply:
Here the certificates are shown in summary. Now that you bring it up, it seems unnecessary. Together with the comment about side-effects in layout generation, maybe it's ok to simplify things and remove certificates and withdrawals from `confirm_transaction` altogether. It was meant mainly as a summary of the things contained in the transaction, but since the user has confirmed them once, why inform him about them again?

Reply:
Discussed offline - certificates and withdrawals will be removed from the last screen.